File: 周赛/第 465 场周赛/q3.py
from typing import List, Tuple, Dict, Set, Optional
from collections import defaultdict, Counter, deque
from itertools import permutations, combinations
from datetime import datetime, date, time, timedelta
from time import time as timestamp, sleep
from functools import cmp_to_key, lru_cache, reduce
from math import gcd, sqrt, log, ceil, floor, inf
from bisect import bisect_left, bisect_right
from heapq import heappush, heappop, heapify, nsmallest, nlargest
from sys import setrecursionlimit, stdin, stdout
setrecursionlimit(5 * 10 ** 5 + 1)
input = lambda: stdin.readline().rstrip()
def _max(a, b): return a if a > b else b
def _min(a, b): return a if a < b else b
MOD = 10 ** 9 + 7

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        w = max(nums).bit_length()
        u = 1 << w
        f = [0] * u  # f[s] 表示能由 s 的子集构成的原始数的最大值
        for x in nums:
            f[x] = x  # 初始值

        for s in range(u):  # 从小到大枚举集合 s
            for i in range(w):  # 枚举 s 中的元素 i
                if s >> i & 1:  # i 属于集合 s
                    v = f[s ^ (1 << i)]  # 从 s 的子集 s \ {i} 转移过来
                    if v > f[s]:
                        f[s] = v  # 手写 max 更快

        return max(x * f[(u - 1) ^ x] for x in nums)
    

# print((1 << 20) * 20) # 20971520

File: 周赛/第 465 场周赛/q3.cpp
#include <vector>
#include <algorithm>
#include <climits>
using namespace std;

class Solution {
public:
    long long maxProduct(vector<int>& nums) {
        // 计算所有数中最大值的二进制位数
        int max_val = *max_element(nums.begin(), nums.end());
        int w = 0;
        int temp = max_val;
        while (temp) {
            w++;
            temp >>= 1;
        }
        // 如果 max_val 是 0，w 会是 0，但至少需要 1 位
        if (max_val == 0) w = 1;

        int u = 1 << w; // 即 2^w
        vector<int> f(u, 0); // f[s] 表示能由 s 的子集构成的原始数的最大值

        // 初始化：每个数字本身
        for (int x : nums) {
            f[x] = x;
        }

        // 标准的 SOS DP：枚举所有集合 s，从小到大
        for (int s = 0; s < u; s++) {
            for (int i = 0; i < w; i++) {
                if (s >> i & 1) { // 如果第 i 位为 1
                    f[s] = max(f[s], f[s ^ (1 << i)]); // 从 s 去掉第 i 位转移
                }
            }
        }

        // 枚举每个数 x，找与其无公共 1 位的最大 f[~x]
        long long ans = 0;
        for (int x : nums) {
            int complement = (u - 1) ^ x; // 取反，限制在 w 位内
            ans = max(ans, (long long)x * f[complement]);
        }

        return ans;
    }
};